master/buildbot/process/build.py

```diff
@@ -225,23 +225,32 @@ def setupPropertiesKnownBeforeBuildStarts(props, requests, builder,
         # get worker properties
         # navigate our way back to the L{buildbot.worker.Worker}
         # object that came from the config, and get its properties
-        workerforbuilder.worker.setupProperties(props)
+        if workerforbuilder is not None:
+            workerforbuilder.worker.setupProperties(props)
 
-    def setupOwnProperties(self):
+    @staticmethod
+    def setupBuildProperties(props, requests, sources=None):
         # now set some properties of our own, corresponding to the
         # build itself
-        props = self.getProperties()
-        props.setProperty("buildnumber", self.number, "Build")
+        props.setProperty("buildnumber", None, "Build")
+
+        if sources is None:
+            sources = requests[0].mergeSourceStampsWith(requests[1:])
 
-        if self.sources and len(self.sources) == 1:
+        if sources and len(sources) == 1:
             # old interface for backwards compatibility
-            source = self.sources[0]
+            source = sources[0]
             props.setProperty("branch", source.branch, "Build")
             props.setProperty("revision", source.revision, "Build")
             props.setProperty("repository", source.repository, "Build")
             props.setProperty("codebase", source.codebase, "Build")
             props.setProperty("project", source.project, "Build")
 
+    def setupOwnProperties(self):
+        props = self.getProperties()
+        Build.setupBuildProperties(props, self.requests, self.sources)
+        props.setProperty("buildnumber", self.number, "Build")
```

Comment:
I think it makes sense to pass the buildnumber as a parameter to setupBuildProperties.